### DISCUSSION TITLE
How to stop a path from entering the middleware function if that path has already gone through the middleware once?

### DISCUSSION THREAD

**Discussion by isas5675** [vercel/next.js#48538] [Help]
### Summary

To put it even simply, we are able to control which paths should enter the middleware function with the help of "matcher" configuration. But how to control the number of times a path enters the middleware function? 

We have configured the matcher for 7 differernt paths. They should enter the middleware function only once. If they have entered once, they should not enter again.

In our case, we want to fetch the location and set it in cookies. If it's already set, then don't hit the middleware again.

The reason we want to achieve this is to stop unnecessary middleware invocations as we are charged by Vercel for every middleware invocation.

Our site is hosted on Vercel.

Please find the middleware code for more clarity.

### Additional information

```js
// middleware.js

import { NextResponse } from "next/server";

export async function middleware(req) {
  const locationFetched = Boolean(req.cookies.get("location-details"));

  if (!locationFetched) {
    console.log(locationFetched);
    console.log("-----------------------");
    const locationDetails = req.geo;

    const response = NextResponse.next();
    response.cookies.set("location-details", JSON.stringify(locationDetails));

    return response;
  } else {
    return NextResponse.next();
  }
}

export const config = {
  matcher: [
    "/:slug?",
    "/",
    "/home",
    "/all-events",
    "/events",
    "/clubs",
    "/deals",
  ],
};
```

**Comment by qpre** [comment]
just checking that it is already set would require a middleware function.

Is cost is really an issue here ? (https://vercel.com/docs/concepts/functions/edge-middleware/usage-and-pricing#invocation-limits) ?

If you do not have cross-domain requirements, you could go with another implementation, like having a key on client side storage (such as localStorage), if this key is set do nothing, it is not use fetch to call an api route that sets the cookie. but this isn't ideal as localStorage and cookie storages might get out of sync.

**Reply by isas5675** [reply]
Yes, cost is a big issue here. Because our website is accessed by hundreds of people everyday. And on the previous Sunday, we had more than 600,000 middleware invocations. And because of that the limits were getting crossed very soon.

Any other idea?

**Reply by qpre** [reply]
Oh i see, thought you were on a team membership with that kind of volume.

**Comment by icyJoseph** [comment]
If I remember correctly, the matcher can't be dynamic. The middleware building process is a bit special, where the matcher object is parsed as an AST, and from there handled by the server side logic. So you can't really affect the matcher behaviour at runtime.

It is also slightly more difficult because of the order in which the middleware kicks in: https://nextjs.org/docs/advanced-features/middleware#matching-paths

Perhaps you could redirect users missing the location to a special route, which sets the location as a cookie, and redirects them back to where they came from? Then you could match that route int he middleware and keep on working as you want.
